### PR TITLE
feat(codepipeline): change to stand-alone Artifacts

### LINKED ISSUE #2230: CodePipeline: artifact names should be restricted to 100 characters
Right now it's very easy for the autogenerated artifact names to become humongously long:

```
Artifact_delivlibtesthuijbersCodeCommitPipelineGenerateTwoArtifactsActionGenerateTwoArtifacts4EDEFC2E
```

This is 101 characters, the limit is 100.

### PR DESCRIPTION
**Pull request by skinny85**
This commit changes the way Artifacts are used in CodePipeline.
Instead of being properties on the Actions,
they are now stand-alone objects,
created independently of the Actions,
and referenced when instantiating them.
This is to not force users to assign Actions to intermediate variables
when defining a Pipeline.

This change had a few interesting consequences:
* We no longer needed the abstract subclasses of Action like SourceAction,
  DeployAction, etc., and so they were removed.
* The old naming convention of `inputArtifacts` and `outputArtifactNames`
  was shortened to be simply `inputs` and `outputs`.
* There was no longer any need to differentiate Build and Test Actions,
  and so the two CodeBuild and Jenkins Actions were merged into one.

BREAKING CHANGE: CodePipeline Actions no longer have the `outputArtifact`
and `outputArtifacts` properties.
* `inputArtifact(s)` and `additionalInputArtifacts` properties were renamed
  to `input(s)` and `extraInputs`.
* `outputArtifactName(s)` and `additionalOutputArtifactNames` properties
  were renamed to `output(s)` and `extraOutputs`.
* The classes `CodeBuildBuildAction` and `CodeBuildTestAction` were merged
  into one class `CodeBuildAction`.
* The classes `JenkinsBuildAction` and `JenkinsTestAction` were merged
  into one class `JenkinsAction`.

Fixes #2230

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-codepipeline/lib/action.ts [hunk 1]
```diff
@@ -27,12 +27,29 @@ export interface ActionArtifactBounds {
...
+    minOutputs: 1,
+    maxOutputs: 1,
+  };
+}
+
+/**
+ * The ArtifactBounds that make sense for deploy Actions -
+ * they have exactly one input, and don't produce any outputs.
+ */
+export function deployArtifactBounds(): ActionArtifactBounds {
```
Comment (aecollver) [inline packages/@aws-cdk/aws-codepipeline/lib/action.ts]: This isn't true (I can create deploy actions with output artifacts) and artifact bounds are set per-action type. Let's minimally move this to aws-codepipeline-actions as a helper function or ideally let each action type declare its own artifact bounds.

Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/action.ts]: Moved these 2 to `aws-codepipeline-actions` and made them private.

File: packages/@aws-cdk/app-delivery/test/test.pipeline-deploy-stack-action.ts [hunk 2]
```diff
@@ -305,7 +304,7 @@ class FakeAction extends codepipeline.Action {
   constructor(actionName: string) {
     super({
       actionName,
-      artifactBounds: codepipeline.defaultBounds(),
+      artifactBounds: { minInputs: 0, maxInputs: 5, minOutputs: 0, maxOutputs: 5 },
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/app-delivery/test/test.pipeline-deploy-stack-action.ts]: It wasn't really useful, and I don't even think the name was accurate (having 0 to 5 inputs & outputs is definitely _not_ a default for Actions).

File: packages/@aws-cdk/aws-codepipeline-actions/lib/codebuild/build-action.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+   * The type of the Action that determines its CodePipeline Category -
+   * Build, or Test.
+   *
+   * @default CodeBuildActionType.BUILD
+   */
+  readonly type?: CodeBuildActionType;
+}
+
+/**
+ * CodePipeline build Action that uses AWS CodeBuild.
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/codebuild/build-action.ts]: Changed.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/codebuild/build-action.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    if (this.actionInputArtifacts.length > 1) {
+      this.configuration.PrimarySource = new cdk.Token(() => this.actionInputArtifacts[0].artifactName);
+    }
+  }
+
+  protected bind(info: codepipeline.ActionBind): void {
+    // grant the Pipeline role the required permissions to this Project
+    info.role.addToPolicy(new iam.PolicyStatement()
+    .addResource(this.props.project.projectArn)
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/codebuild/build-action.ts]: Done.

File: packages/@aws-cdk/aws-codepipeline/lib/artifact.ts [hunk 5]
```diff
@@ -1,11 +1,29 @@
...
 import { Token } from "@aws-cdk/cdk";
+import validation = require('./validation');
 
 /**
  * An output artifact of an action. Artifacts can be used as input by some actions.
  */
 export class Artifact {
-  constructor(readonly artifactName: string) {
+  /**
+   *
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/artifact.ts]: Added.

File: packages/@aws-cdk/aws-codepipeline/lib/artifact.ts [hunk 6]
```diff
@@ -72,12 +99,18 @@ export class Artifact {
...
+    return new ArtifactPath(Artifact.artifact(artifactName), fileName);
+  }
+
   constructor(readonly artifact: Artifact, readonly fileName: string) {
 
   }
 
   get location() {
-    return `${this.artifact.artifactName}::${this.fileName}`;
+    return `${this.artifact.artifactName
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/artifact.ts]: Done.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 7]
```diff
@@ -445,6 +447,52 @@ export class Pipeline extends Resource implements IPipeline {
...
+  private validateArtifacts(): string[] {
+    const ret = new Array<string>();
+
+    const outputArtifactNames = new Set<string>();
+    for (const stage of this.stages) {
+      const sortedActions = stage.actions.sort((a1, a2) => a1.runOrder - a2.runOrder);
+
+      // start with inputs
+      for (const action of sortedActions) {
+        // TODO: remove "as any"
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Done.

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts [hunk 8]
```diff
@@ -45,6 +46,23 @@ export class Stage implements IStage {
...
+      const outputArtifacts: Artifact[] = (action as any).actionOutputArtifacts;
+
+      const unnamedOutputs = outputArtifacts.filter(o => !o.artifactName);
+
+      for (const outputArtifact of outputArtifacts) {
+        if (!outputArtifact.artifactName) {
+          const artifactName = `Artifact_${this.stageName}_${action.actionName}` + (unnamedOutputs.length === 1
+            ? ''
+            : '_' + (unnamedOutputs.indexOf(outputArtifact) + 1));
+          (outputArtifact as any)._setName(artifactName);
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Removed `as any`.

File: packages/@aws-cdk/aws-codepipeline/lib/artifact.ts [hunk 9]
```diff
@@ -1,11 +1,29 @@
...
+   *
+   * @param name
+   */
+  public static artifact(name: string): Artifact {
+    return new Artifact(name);
+  }
+
+  private _artifactName?: string;
+
+  constructor(artifactName?: string) {
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/artifact.ts]: Can't - it's used in `Artifact`.

[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline/lib/artifact.ts]: ?! It's used in the static method, no?

```ts
class Foo {
  public static makeFoo(): Foo { return new Foo() ; }
  private constructor() { }
}
```

Now only `Foo.makeFoo` can be used to make foo and not `new Foo()`.

**Comment by skinny85** [comment]
Changes after review comments.

**Comment by skinny85** [comment]
Added a missing test for action Artifacts bound validation.

File: packages/@aws-cdk/aws-codepipeline/lib/artifact.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/artifact.ts]: Right, except the static method is not that simple in this case:

```ts
  public static artifactPath(artifactName: string, fileName: string): ArtifactPath {
    return new ArtifactPath(Artifact.artifact(artifactName), fileName);
  }
```

So it's not simply equivalent to `new ArtifactPath(artifact, fileName)`.

[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline/lib/artifact.ts]: I am taking about the ctor of `Artifact`... I feel I am missing something... But nevermind, I don't really care

**Pull request merged**